### PR TITLE
yoav - Auto Calibration (23 files) (Don't merge)

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 1]
```diff
@@ -26,6 +27,7 @@
 public class Swerve extends GBSubsystem {
 
     public final Lock odometryLock;
+    private ProfiledPIDController kp;
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:30]: why is a pidController called kp?

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 2]
```diff
@@ -41,6 +43,15 @@ public Swerve() {
 
         configurePathPlanner();
 
+        kp = new ProfiledPIDController(1, 0, 0, SwerveConstants.ROTATION_CONSTRAINTS);
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:46]: magic

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 3]
```diff
@@ -41,6 +43,15 @@ public Swerve() {
...
+    }
+
+    public double getKp(){
+        return kp.getP();
+    }
+
+    public void setKP(double kp){
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:53]:
```suggestion
    public void setKp(double kp){
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java [hunk 4]
```diff
@@ -4,24 +4,51 @@
...
+                                new Pair<>(90.0, -90.0),
+                                () -> SWERVE.getNonAbsoluteHeading().getDegrees(),
+                                SWERVE::getKp,
+                                position -> SWERVE.rotateToAngle(Rotation2d.fromDegrees(position)),
+                                SWERVE::setKP,
+                                targetAngle -> SWERVE.isAtAngle(Rotation2d.fromDegrees(targetAngle)),
+                                SWERVE::stop
+                        )
+                );
+    }
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:49]: 🧙‍♂️🧙‍♂️

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/FindHighestP.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     */
+
+    public FindHighestP(
+            boolean isSetControlNeedToRunPeriodic,
+            double tolerance, double timeoutForActionSeconds, double errorToKpValueFactor, double multiPFactor,
+            Pair<Double, Double> valuesToRunFor,
+            DoubleSupplier currentValueSupplier, DoubleSupplier currentKpValueSupplier,
+            Consumer<Double> setControl, Consumer<Double> setKp,
+            Predicate<Double> isAtPose,
+            Runnable stopAtEnd
```
Comment (YoniKiriaty) [inline src/main/java/frc/utils/calibration/autocalibration/kpfinding/FindHighestP.java:41]: maybe make a config class for all this stuff where you can set these variables?
this looks like too many paramaters

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        isExe = false;
+    }
+
+    protected void replaceTargetValue() {
+        usedTargetValue = (usedTargetValue == valuesToRunFor.getFirst()) ? valuesToRunFor.getSecond() : valuesToRunFor.getFirst();
+    }
+
+    protected boolean hasOscillated(double currentPosition) {
+        return !((valuesToRunFor.getFirst() + tolerance >= currentPosition && currentPosition >= valuesToRunFor.getSecond() - tolerance)
+                || (valuesToRunFor.getFirst() - tolerance <= currentPosition && currentPosition <= valuesToRunFor.getSecond() + tolerance));
```
Comment (YoniKiriaty) [inline src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java:89]: use the `Math.abs()` way of tolerance

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java
Comment (YoniKiriaty) [inline src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java:1]: i suggest you look into RepeatCommand or some other command structure

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:30]: will be deleted

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 2 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:46]: will be deleted

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/FindHighestP.java (hunk 5 above)
Comment (YHerm) [reply src/main/java/frc/utils/calibration/autocalibration/kpfinding/FindHighestP.java:41]: I don't see a good reason for that to be honest

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 3 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:53]: will be deleted

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java (hunk 6 above)
Comment (YHerm) [reply src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java:89]: i can't do abs, it is between two points and given value

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java
Comment (YHerm) [reply src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java:1]: i looked but i dont think repeat command can help me a lot here, all the ways i can think of them to use that is worse then what there is now in my opinion

File: src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java (hunk 6 above)
Comment (YoniKiriaty) [reply src/main/java/frc/utils/calibration/autocalibration/kpfinding/PFinding.java]: then split it into some boolean variables
like `boolean firstCondition = (valuesToRunFor.getFirst() + tolerance >= currentPosition && currentPosition >= valuesToRunFor.getSecond() - tolerance);
boolean secondCondition = (valuesToRunFor.getFirst() - tolerance <= currentPosition && currentPosition <= valuesToRunFor.getSecond() + tolerance);
return !(firstCondition || secondCondition)`

**Pull request closed without merging**